src/plugins/arithmetic.ts

```diff
@@ -283,3 +284,126 @@ export function mulOrDivToBitShift(fromMul = true, fromDiv = true): Plugin {
 }
 
 export const bitShiftPlugins = [bitShiftToMulOrDiv(), mulOrDivToBitShift()];
+
+export type IntDecomposition = [
+  // [k,b,e,d,cost] represents a value k * pow(b,e) + d, cost = lg(k)+lg(b)+lg(d)
+  bigint,
+  bigint,
+  bigint,
+  bigint,
+  number
+];
+
+// assert (1000 ≤ |n|)
+export function decomposeInt(n: bigint): IntDecomposition[] {
+  return decomposeAnyInt(n, n);
+}
+
+// assert (1000 ≤ x ≤ y || x ≤ y ≤ -1000)
+export function decomposeAnyInt(x: bigint, y: bigint): IntDecomposition[] {
+  return x > 0
+    ? _decomposeAnyInt(x, y)
+    : _decomposeAnyInt(-y, -x).map(([k, b, e, d, c]) => [-k, b, e, -d, c]);
+}
+
+function lg(n: bigint): number {
+  if (n < 0) n = -n;
+  return n > 1000000000000
+    ? 2 + n.toString(16).length
+    : n > 99
+    ? n.toString().length
+    : n > 9
+    ? 2
+    : 1;
+}
+
+function betterOrEqual(
+  [k1, b1, e1, d1, c1]: IntDecomposition,
+  [k2, b2, e2, d2, c2]: IntDecomposition
+): boolean {
+  return (
+    (b1 === 2n || b2 !== 2n) &&
+    (b1 === 10n || b2 !== 10n) &&
+    (k1 === 1n || k2 !== 1n) &&
+    (d1 === 0n || d2 !== 0n) &&
+    c1 <= c2
+  );
+}
+
+function ceilDiv(a: bigint, b: bigint) {
+  return (a + (b - 1n)) / b;
+}
+
+// assert (1000 ≤ x ≤ y)
+// Find decompositions x ≤ k * b^e + d ≤ y s.t. 1 ≤ k < 100, 2 ≤ b ≤ 10, |d| < 100
+function _decomposeAnyInt(x: bigint, y: bigint): IntDecomposition[] {
+  const xd = x - 99n;
+  const yd = y + 99n;
+  const decompositions: IntDecomposition[] = [];
+  for (let b = 2n; b <= 10n; b++) {
+    for (
+      let e = 2n, be = b * b, kx = ceilDiv(xd, be), ky = yd / be;
+      kx <= ky;
+      e++, be *= b, kx = ceilDiv(kx, b), ky /= b
+    ) {
+      if (be < 1000) continue;
+      const kmax = ky > kx ? kx + 1n : kx;
+      for (let k = kx; k <= kmax; k++) {
```

Comment:
```suggestion
      for (let k = kx; k <= kmax; k++) {
        if (k % b === 0n && lg(e) == lg(e + 1n)) continue;
```

- This makes the algo slightly faster: 13 ms → 4 ms for `decomposeInt(10n**1000n)`.
- `decomposeInt(3n<<20n)` generates 3&times;2<sup>20</sup> instead of 6&times;2<sup>19</sup>.